doc/man1/mkethcon.1
.TH MKETHCON 1 "2008-08-25" 1.0 "LINX"
.SH NAME
mkethcon \- create LINX Ethernet connections.
.SH SYNOPSIS
.B mkethcon [OPTIONS] <connection>

.SH DESCRIPTION
Create a LINX Ethernet connection to a remote node.

.SH OPTIONS

.IP "-h, --help"
Display help and exit.

.IP "Mandatory options"

.IP "-m, --mac=MAC address"
Specify the remote MAC address of the node to connect to, e.g.
0a:1b:2c:3d:4d:5e.

.IP "-i, --if=interface"
Specify the Ethernet interface to use, e.g. eth0.

.IP "Optional options"

.IP "--window_size=nnn"
This is the send/receive window_size, in number of packets on the link,
and may need to be modified to adapt LINX to really slow or really fast
Ethernet performance.
.br
Both the local and remote side negotiate the window size, being the
smallest of the two sizes the selected one.
The default 128 messages should be sufficient in most configurations.
The window size shall always be of a power of 2.
Size 0 means to use the default window size.

.IP "--defer_queue_size=nnn"
The defer queue size is in packages, with the size depending on the LINX link.
The defer queue is used on the sender side when the send queue is full.
Every message sent, when the send queue is full, is stored in the defer queue,
until the send queue has room for more messages.
The defer queue size is the maximum number of packages, with messages,
that can be stored in the defer queue before the link is disconnected,
due to lack or resources.
The default value 2048 packages should be sufficient in most systems.
Size 0 means to use the default defer queue size.

.IP "--conn_tmo=nnn"
The connect timeout specifies the time, in milliseconds, to wait until a
connection attempt fails and a new attempt is made. Once connected, it is used
as a supervision time-out, i.e. every conn_tmo/3 milliseconds, LINX checks if
any packets from the peer has been received. LINX disconnects the connection
after four consequtive conn_tmo/3 timeout periods without any packets from the
peer. Tmo 0 means to use the default timeout (1000ms).

.IP "--coreid=nn"
The coreid is used in a multicore environment to configure the destination 
coreid.

.IP "--mtu=nnn"
The MTU (Maximum Transmission Unit) specifies the size in  bytes of the
largets packet that the Ethernet protocol can pass onwards (this excludes the
size of the Ethernet header). If not set the MTU is fetched from the interface.
Typcially a MTU of 1500 bytes is used for Ethernet.

.SH FILES
None.

.SH DIAGNOSTICS
None.

.SH KNOWN BUGS
None.

.SH EXAMPLES
.nf
mkethcon --mac=01:23:a4:4f:b3:ac --if=eth0 ethcon_A
mkethcon -m 01:23:a4:4f:b3:ac -i eth0 ethcon_A
.fi

.SH "SEE ALSO"
.BR linx "(7), "
.BR linxcfg "(1), "
.BR linxstat "(1), "
.BR linxdisc "(8), "
.BR mklink "(1), "
.BR rmethcon "(1)"

.SH AUTHOR
Enea LINX team

.SH COPYRIGHT
Copyright (c) 2006-2007, Enea Software AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
Neither the name of Enea Software AB nor the names of its
contributors may be used to endorse or promote products derived from this
software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


ChangeLog
Version 2.5.1 (2011-07-08)
==========
* The limit of number of sockets per link allowed lowered from 1024 to two.
* Removed a check in LINX kernel module for max signal size.
* Fixes for the multicore header in the ECM and Wireshark patch.
		
Version 2.5.0 (2011-04-26)
==========
* Fixed a bug in linxstat.
* Updated manpage for mkethcon and help info for mkethcon and linxcfg to reflect
  reality.
* libcfg was updated to support cmcl connections.
* Fixed minor bugs in rm*con binaries.
* Fixed build errors for various kernel versions. ipv6 for <2.6.22 and db.c for
  newer ones >2.6.36.
* Fixed some issues in the linx Gateway related to LINXCR-1397
* Added support for ipv6 in the tcp connection manager.
* Improved the performance of intra-process communication with pointer passing.
* Fixed some issues found by coverity.
* Removed some warnings when building linxstat.
* Remove all BUG BUG_ON WARN_ON from linx modules to avoid using __FILE__
  define in the errors. This define causes the linx module to be different
  if compiled in different folder.
* Added fixes in the ECM were a malformed received packet could lead to
  kernel panic.
* Fixed issue with multi-access to LINX sockets from user-space
  applications could lead to kernel panic.
* Added new features to linxstat: print info only for a given huntname
  and display only a specified number of characters from huntname
* Added coreid header
* Added support for 2.5.35 kernels.
* Fixed bug with negative signal filter.
* Fixed bug with missing internal signals in the message trace.
* Added patch file for Ethernet CM to compensate for faulty Ethernet
  drivers that don't allocate enough headroom in skb when using VLAN.
* Added support for 2.6.33 and 2.6.34 kernels.
* Fixed vlan bug reproducing on certain machines
* Fixed bug with receive_w_tmo where tmo=0 resulted in no signal above
  4095 could be received.
* Fixed string handling bug in linxgws.
* Fixed memory allocation issue in linxgws.
* Fixed a bug in linxdisc regarding several interfaces.
* Added ip config in linxnet conf file
* Linxstat now uses a function interface to retrieve info instead of
ioctl.
	
Version 2.4.0 (2010-02-22)
==========	
* Added missing riocm.txt to MANIFEST
* Added RIOCM documentation. net/linx/riocm/riocm.txt
* Updated RELEASE_NOTES for 2.4.0 release.
* Added support for creating links using RapidIO CM connections in linxcfg.

Version 2.4.0-beta (2010-01-15)
==========
* Added a RapidIO connection manager.
* Remove debug printk in device notification handler.

Version 2.3.0 (2009-12-9)
==========
* Fixed build errors for vanilla kernel 2.6.32
* Example conf file for Linxdisc now emphasizes that VLAN interfaces are supported.
* RLNH control data is now always sent from kernel context.
* Altered the keepalive mechanism for the Ethernet CM. The old implementation
  was sensitive to packet loss. Contributed by paull7 @ sourceforge. Note: The
  new implementation is backwards compatible.
* Added support for 2.6.31 kernels.
* Fix LINX_IOCTL_INFO ioctl request for gathering process' receive filters.
* Document LINX_INFO_RECV_QUEUE_2.
* Drop tcp connect packets that doesn't match the current protocol
version.
* Add hardware address size in ethernet connection header.
* Man page for linx_request_new_link and linx_cancel_new_link was missing
* Fixed bug where the request ack bit was not set when the sliding window queue
  is becoming full.
* Fixed memory leaks in the code for create and destroy connections and links.	
* New shared memory connection manager.
* Wireshark plugin now supports LINX TCP Connection manager.
* Wrong RLNH protocol version was listed in documentation.	
	
Version 2.2.0 (2009-03-31)
==========
* New Linxdisc and Linx internal messages patches for Wireshark 1.0.6
* New LINX patches for tcpdump 4.0.0 and libpcap 1.0.0
* Trace of internal LINX messages now use official LINX ethernet type, 0x8911.
* Added new function linx_get_owner(), return the pid of the owner of a LINX endpoint.	
* Added missing files to MANIFEST.
* Example files now have the correct copyright header.
* Fixed oob-issue where the a link in some cases wasn't created.
* Removed a state in the tcp cm state machine.
* Did some major cleanup in tcp cm.
* Removed a kernel function from af_linx.c
* Fixed a bug in tcp cm regarding halt of workqueue when connecting to dead peers
* Linux Gateway Server and Client added to distribution, also gwcmd tool to test Linx Gateway Servers. 
* Removed bug_on's and fixed better error reporting back to user space on failed send.
* Fix bug in linx_info_filter_32
* Fixed alignment issue in linxstat.
* Removed linx_create/destroy_eth/tcp_link functions.
* Avoid calling kmalloc(0) from Linx IPC-layer.

Version 2.1.0 (2008-11-10)
==========	
* LINX now runs on x86 2.6.27 Linux kernel.
* Add support for mips(64).
* Add patches to libpcap and tcpdump configure scripts for Octeon.
* The TCP CM now support OOB messaging.
* Fixes for running w errorchecks enabled.
* Improved garbage collection of sockets in the tcp_cm.
* Minor changes of linxstat output.
* The linxstat program can now tell if a signal in the receive queue is OOB.
* Support for allocating fragmentated skb's is now a LINX kernel module param and no longer
  a comple-time option.
* Fixed bug when building the LINX kernel module with memory fragmentation support on and
  sending large unpaged signals.
* Fixed bug in LINX message trace. If a fragmented message was received, only the first
  fragment was traced.
* Added optimized linx_do_sendmsg call for attach/hunt/tmo/newlink signals.
* Pulled back warning level a bit to avoid complaints from kernel headers.
* Fixed bug that lead to the LINX kernel module couldn't be unloaded.
* linx_skb_copy_bits removed
* linx_ioctl_info_*() functions moved to a seperate file.
* Internal handling of attach/tmo/newlink signals is simplified.
* Updated liblinx to user LINX_IOCTL_SEND and LINX_IOCTL_RECEIVE instead of sendto/recvfrom/sendmsg

linxcfg/mkethcon.c
/*
 * Copyright (c) 2006-2007, Enea Software AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of Enea Software AB nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * File: mkethcon.c - Command to make a Linx Ethernet connection.
 */
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <linx.h>
#include <linxcfg.h>
#include "db_utils.h"
#include "ethcm_utils.h"

#define OPT_WINDOW_SIZE 256
#define OPT_DEFER_QUEUE_SIZE 257
#define OPT_SEND_TMO 258
#define OPT_NACK_TMO 259
#define OPT_CONN_TMO 260
#define OPT_LIVE_TMO 261
#define OPT_MTU 262
#define OPT_COREID 263

struct option long_opts[] = {
        {"mac", 1, NULL, (int)'m'},
        {"if", 1, NULL, (int)'i'},
        {"window_size", 1, NULL, OPT_WINDOW_SIZE},
        {"defer_queue_size", 1, NULL, OPT_DEFER_QUEUE_SIZE},
        {"send_tmo", 1, NULL, OPT_SEND_TMO},
        {"nack_tmo", 1, NULL, OPT_NACK_TMO},
        {"conn_tmo", 1, NULL, OPT_CONN_TMO},
        {"live_tmo", 1, NULL, OPT_LIVE_TMO},
        {"mtu", 1, NULL, OPT_MTU},
        {"coreid", 1, NULL, OPT_COREID},
        {"help", 0, NULL, (int)'h'},
        {"silent", 0, NULL, (int)'s'},
        {NULL, 0, NULL, 0}
};

static char *usage =
"mkethcon [OPTIONS] <connection>\n"
"Create a Linx Ethernet connection\n"
"OPTIONS:\n"
"    -m, --mac=<mac address>\n"
"        Specify the remote address to connect to,\n"
"        e.g. 0a:1b:2c:3d:4d:5e.\n"
"        Mandatory option.\n"
"\n"
"    -i, --if=<interface>\n"
"        Specify the Ethernet interface to use, e.g. eth0.\n"
"        Mandatory option.\n"
"\n"
"    --mtu=<size>\n"
"        The MTU (Maximum Transmission Unit) specifies the size\n"
"        in bytes of the largets packet that the Ethernet protocol\n"
"        can pass onwards (this excludes the size of the Ethernet\n"
"        header). If not set the MTU is fetched from the interface.\n"
"        Typcially a MTU of 1500 bytes is used for Ethernet.\n"
"\n"
"    --window_size=<size>\n"
"        The send and receive window_size may need to be \n"
"        modified to adapt linx to really slow or really fast\n"
"        ethernet performance, the default of 128 messages\n"
"        should be sufficient in most configurations. The\n"
"        window size shall always be of a power of 2\n"
"        Size 0 means default window size\n"
"\n"
"    --defer_queue_size=<size>\n"
"        The defer queue is used on the sender side When the\n"
"        send window is full, every message sent when the send\n"
"        window is full is stored in the defer queue until the\n"
"        send window has room for more messages. The defer\n"
"        queue size specifies the maximum number of messages\n"
"        that can be stored in the defer queue before the link\n"
"        is disconnected due to lack or resources. The default\n"
"        value of 2048 should be sufficient in most systems.\n"
"        Size 0 means default defer queue size\n"
"\n"
"    --conn_tmo=<tmo>\n"
"        The connect timeout specifies the time, in milliseconds,\n"
"        to wait until a connection attempt fails and a new\n"
"        attempt is made. Once connected, it is used as a \n"
"        supervision time-out, i.e. every conn_tmo/3 milliseconds,\n"
"        LINX checks if any packets from the peer has been\n"
"        received. LINX disconnects the connection after four\n"
"        consequtive conn_tmo/3 timeout periods without any\n"
"        packets from the peer. Tmo 0 means to use the default\n"
"        timeout (1000ms).\n"
"\n"
"     --coreid=<id>\n"
"        The coreid is used in a multicore environment \n"
"        to configure the destination coreid.\n"
"\n"
"Examples:\n"
"    mkethcon --mac=01:23:a4:4f:b3:ac --if=eth0 ethcon_A\n"
"    mkethcon -m 01:23:a4:4f:b3:ac -i eth0 ethcon_A\n";

static uint32_t get_u32(const char *opt, const char *arg)
{        
        char *end;
        uint32_t u32;

        u32 = (uint32_t)strtoul(arg, &end, 0);
        if (*end != '\0') {
                printf("mkethcon: error in %s\n"
                       "Try 'mkethcon --help' for more information\n",
                       opt);
                exit(1);
        }
        return u32;
}

static int option_supported(int ver, const char *opt)
{
        int major, minor, patch;

        (void)opt; /* Not used at the moment */

        major = (ver >> 24) & 0xff;
        minor = (ver >> 8) & 0xffff;
        patch = ver & 0xff;

        if (major > 2)
                return 0; /* No support */
        if (major == 2 && minor > 1)
                return 0; /* No support */
        
        return 1;
}

int main(int argc, char *argv[])
{
        int c, silent, status, ver;
        char *s, version[14];
        union linx_con_arg con;

        memset(&con.eth, 0, sizeof(con.eth));
        silent = 0;
        con.eth.coreid = -1;
        ver = linx_get_version(version);
        if (ver == -1) {
                printf("WARNING: failed to retrieve LINX version\n");
                ver = 0;
                strcpy(version, "unknown");
        }

        for (c = 0; c != -1;) {
                c = getopt_long(argc, argv, "hi:m:s", long_opts, NULL);
                switch (c) {
                case 'h':
                        printf("Usage: %s\n", usage);
                        return 0;
                case 's':
                        silent = 1;
                        break;
                case 'm':
                        con.eth.mac = optarg;
                        break;
                case 'i':
                        con.eth.ethif = optarg;
                        break;
                case OPT_MTU:
                        con.eth.mtu = get_u32("--mtu", optarg);
                        break;
                case OPT_WINDOW_SIZE:
                        con.eth.window_size = get_u32("--window_size", optarg);
                        if ((con.eth.window_size & (con.eth.window_size - 1)) != 0) {
                                printf("mkethcon: error in --window_size: "
                                       "Not power of 2\n"
                                       "Try 'mkethcon --help' for more information\n");
                                return 1;
                        }
                        break;
                case OPT_DEFER_QUEUE_SIZE:
                        con.eth.defer_queue_size = get_u32("--defer_queue_size", optarg);
                        break;
                case OPT_SEND_TMO:
                        con.eth.send_tmo = get_u32("--send_tmo", optarg);
                        if (!option_supported(ver, "--send_tmo")) {
                                printf("WARNING! --send_tmo is not supported in "
                                       "LINX %s. Option is ignored.\n", version);
                        }
                        break;
                case OPT_NACK_TMO:
                        con.eth.nack_tmo = get_u32("--nack_tmo", optarg);
                        if (!option_supported(ver, "--nack_tmo")) {
                                printf("WARNING! --nack_tmo is not supported in "
                                       "LINX %s. Option is ignored.\n", version);
                        }
                        break;
                case OPT_CONN_TMO:
                        con.eth.conn_tmo = get_u32("--conn_tmo", optarg);
                        break;
                case OPT_LIVE_TMO:
                        con.eth.live_tmo = get_u32("--live_tmo", optarg);
                        if (!option_supported(ver, "--live_tmo")) {
                                printf("WARNING! --live_tmo is not supported in "
                                       "LINX %s. Option is ignored.\n", version);
                        }
                        break;
                case OPT_COREID:
                       con.eth.coreid = get_u32("--coreid", optarg);
                       break;
                case '?':
                        return 1;
                }
        }
        con.eth.name = argv[optind];
        con.eth.features = "";

        if (con.eth.name == NULL) {
                printf("mkethcon: no connection specified\n"
                       "Try 'mkethcon --help' for more information\n");
                return 1;
        }
        if (con.eth.mac == NULL) {
                printf("mkethcon: -m, --mac is missing\n"
                       "Try 'mkethcon --help' for more information\n");
                return 1;
        }
        if (con.eth.ethif == NULL) {
                printf("mkethcon: -i, --if is missing\n"
                       "Try 'mkethcon --help' for more information\n");
                return 1;
        }

        s = NULL;
        status = linx_create_connection(LINX_CON_ETH, &con, &s);
        if (status != 0) {
                printf("mkethcon: couldn't create connection '%s': %s\n",
                       con.eth.name, strerror(status));
                return 1;
        }
        
        if (!silent)
                printf("mkethcon: created connection '%s'.\n"
                       "Now use 'mklink -c %s ...' to create a link.\n",
                       s, s);
        free(s);
        return 0;
}
